lib/notifications/notification.ex

```diff
@@ -92,61 +98,101 @@ defmodule Notifications.Notification do
           Skate.Repo.one!(
             from(n in DbNotification,
               join: bw in assoc(n, :block_waiver),
-              select: %{id: n.id, created_at: n.created_at},
-              select_merge: %{
-                reason: bw.reason,
-                route_ids: bw.route_ids,
-                run_ids: bw.run_ids,
-                trip_ids: bw.trip_ids,
-                operator_id: bw.operator_id,
-                operator_name: bw.operator_name,
-                route_id_at_creation: bw.route_id_at_creation,
-                block_id: bw.block_id,
-                service_id: bw.service_id,
-                start_time: bw.start_time,
-                end_time: bw.end_time
-              },
               where:
-                bw.start_time == ^notification_without_id.start_time and
-                  bw.end_time == ^notification_without_id.end_time and
-                  bw.block_id == ^notification_without_id.block_id and
-                  bw.service_id == ^notification_without_id.service_id and
-                  bw.reason == ^notification_without_id.reason
+                bw.start_time == ^block_waiver_values.start_time and
+                  bw.end_time == ^block_waiver_values.end_time and
+                  bw.block_id == ^block_waiver_values.block_id and
+                  bw.service_id == ^block_waiver_values.service_id and
+                  bw.reason == ^block_waiver_values.reason
+            )
+          )
+      end
+
+    struct!(__MODULE__, Map.merge(notification_values, %{id: db_record.id}))
+  end
+
+  def get_or_create_from_bridge_movement(bridge_movement_values) do
+    created_at = DateTime.utc_now() |> DateTime.to_unix()
+
+    {:ok, {source, db_record}} =
+      Skate.Repo.transaction(fn ->
+        # The bridge API doesn't have anything like an "updated at" timestamp
+        # so we assume that, if we see two bridge movements within a blackout
+        # period, they are actually the same movement and what's happening is
+        # that we have multiple servers trying to insert a movement at once.
+        cutoff_time = NaiveDateTime.utc_now() |> NaiveDateTime.add(-@bridge_lowering_blackout)
+
+        existing_record =
+          Skate.Repo.one(
+            from(n in DbNotification,
+              join: bm in assoc(n, :bridge_movement),
+              where: bm.inserted_at > ^cutoff_time,
+              order_by: [desc: bm.inserted_at],
+              limit: 1
             )
           )
+
+        if existing_record do
+          {:existing_record, existing_record}
+        else
+          changeset =
+            DbNotification.bridge_movement_changeset(
+              %DbNotification{created_at: created_at},
+              bridge_movement_values
+            )
+
+          {:ok, new_record} = Skate.Repo.insert(changeset)
```

Comment:
Isn't there still a race condition here? The result of the query done above in `Skate.Repo.one` could have changed, and if I'm reading correctly there's no unique index on the `bridge_movements` table.

Reply:
There's no unique index on `bridge_movements` because we don't have anything that we can actually index on, at least for the `Lowered` status. You might think that the bridge API would say when the bridge was last lowered: it does not. You might think that it would include a `Last-Modified` header: it does not. Hence the hack referred to in the comments.

I should be able to avoid the race condition by increasing the isolation level of the transaction, though that feels kind of like using a sledgehammer to swat a fly. But it should work.

Reply:
Right, sorry, I didn't mean that there ought to be a unique index on `bridge_movements`, just pointing out that there wasn't one like with `block_waivers`.

I think it _might_ be possible to avoid increasing the isolation by doing some complicated stuff where you have an `insert` that pulls its data from a `select [data you want to add] where not exists [statement selecting any potential duplicate that you want to avoid]`. I'm not sure if one of those is actually any more performant than the other. Changing the isolation level is at least easier.

Reply:
Oh, yeah, I'm just blowing off some of my frustration with the limitations of the bridge API. Upping the serialization level shouldn't have any performance impact to speak of at the scale we're doing things at.